sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java

```diff
@@ -0,0 +1,435 @@
+     * @see <a href="https://github.com/opencontainers/image-spec/blob/main/manifest.md">Oci Manifest Specification</a>
+     *
+     * @param data The manifest that needs to be uploaded.
+     * @return The rest response containing the operation result.
+     * @throws ClientAuthenticationException thrown if the client's credentials do not have access to modify the namespace.
+     * @throws NullPointerException thrown if the {@code data} is null.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<UploadManifestResult>> uploadManifestWithResponse(BinaryData data) {
+        if (data == null) {
+            return monoError(logger, new NullPointerException("'data' can't be null."));
+        }
+
+        return withContext(context -> this.uploadManifestWithResponse(data.toByteBuffer(), context));
+    }
+
+    Mono<Response<UploadManifestResult>> uploadManifestWithResponse(ByteBuffer data, Context context) {
+        if (data == null) {
+            return monoError(logger, new NullPointerException("'data' can't be null."));
+        }
+
+        String digest = UtilsImpl.computeDigest(data);
+        return this.registriesImpl.createManifestWithResponseAsync(
+            repositoryName,
+            digest,
+            Flux.just(data),
+            data.remaining(),
+            UtilsImpl.OCI_MANIFEST_MEDIA_TYPE,
+            context).flatMap(response -> {
+                Response<UploadManifestResult> res = new ResponseBase<ContainerRegistriesCreateManifestHeaders, UploadManifestResult>(
+                    response.getRequest(),
+                    response.getStatusCode(),
+                    response.getHeaders(),
+                    new UploadManifestResult(response.getDeserializedHeaders().getDockerContentDigest()),
+                    response.getDeserializedHeaders());
+                return Mono.just(res);
+            }).onErrorMap(UtilsImpl::mapException);
+    }
+
+    /**
+     * Uploads a blob to the repository.
+     * The client currently uploads the entire blob\layer as a single unit.
+     * <p>
+     * The blob is read into memory and then an upload operation is performed as a single operation.
+     * We currently do not support breaking the layer into multiple chunks and uploading them one at a time
+     *
+     * @param data The blob\image content that needs to be uploaded.
+     * @return The operation result.
+     * @throws ClientAuthenticationException thrown if the client's credentials do not have access to modify the namespace.
+     * @throws NullPointerException thrown if the {@code data} is null.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<UploadBlobResult> uploadBlob(BinaryData data) {
+        if (data == null) {
+            return monoError(logger, new NullPointerException("'data' can't be null."));
+        }
+
+        return withContext(context -> this.uploadBlobWithResponse(data.toByteBuffer(), context)).flatMap(FluxUtil::toMono);
```

Comment:
Will the blobs ever be greater than 2GB? If so, this method will fail as ByteBuffer has a 2GB limit

Reply:
Ideally there is no limit on the size of an image. However, since we are reading everything in memory, if we are at a point where we need to support 2GB and above we will have to move to streaming APIs instead.